### PR TITLE
fix(auth)!: Fetch Auth Session offline behavior

### PR DESCRIPTION
**Pull request by Jordan-Nelson**
*Issue #, if available:* https://github.com/aws-amplify/amplify-flutter/issues/2398, https://github.com/aws-amplify/amplify-flutter/issues/760

*Description of changes:*
- Remove `getAWSCredentials` from Fetch Auth Session API
- Add new types, AWSResult and AuthResult, to hold the result of an operation that may have thrown.
- Update exception handling in fetch auth session state machine
- Update `CognitoAuthSession` members to type AuthResult
- Add Fetch Auth Session test cases 
- Update Authenticator to catch NetworkException.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/auth/amplify_auth_cognito_test/test/state/fetch_auth_session_state_machine_test.dart [hunk 1]
```diff
@@ -29,658 +84,943 @@ void main() {
...
+            setUp(() async {
+              await configureAmplify(config);
+              stateMachine.addInstance<CognitoIdentityProviderClient>(
+                MockCognitoIdentityProviderClient(
+                  initiateAuth: expectAsync0(
+                    () async => InitiateAuthResponse(
+                      authenticationResult: AuthenticationResultType(
+                        accessToken: newAccessToken.raw,
+                        refreshToken: refreshToken,
+                        idToken: idToken.raw,
```
Comment (dnys1) [inline packages/auth/amplify_auth_cognito_test/test/state/fetch_auth_session_state_machine_test.dart:349]: Same comment

File: packages/auth/amplify_auth_cognito_dart/lib/src/model/session/cognito_auth_session_result.dart [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import 'package:amplify_core/amplify_core.dart';
+
+/// {@template amplify_auth_cognito.model.cognito_auth_session_result}
+/// Wraps a session value.
+/// {@endtemplate}
+///
+/// The result may contain an error even though the method
+/// fetching it had a success callback.
+class CognitoAuthSessionResult<T>
+    with AWSEquatable<CognitoAuthSessionResult<T>> {
```
Comment (dnys1) [inline packages/auth/amplify_auth_cognito_dart/lib/src/model/session/cognito_auth_session_result.dart:13]: This can be AWSDebuggable, too - only values which are also AWSDebuggable will print, so it's safe

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/fetch_auth_session_state_machine.dart [hunk 3]
```diff
@@ -279,27 +304,66 @@ class FetchAuthSessionStateMachine extends FetchAuthSessionStateMachineBase {
...
+              idToken == null ? null : _FederatedIdentity.cognito(idToken),
+        );
+        identityId = awsCredentialsResult.identityId;
+        awsCredentials = awsCredentialsResult.awsCredentials;
+        credentialsResult = CognitoAuthSessionResult.success(awsCredentials);
+        identityIdResult = CognitoAuthSessionResult.success(identityId);
+      } on Object catch (e) {
+        credentialsResult = CognitoAuthSessionResult.error(e);
+        // return existing identityId if refreshing credentials fails
+        if (identityId != null) {
```
Comment (dnys1) [inline packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/fetch_auth_session_state_machine.dart:340]: This seems to be accomplishing the same as `previousValue`

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/fetch_auth_session_state_machine.dart [hunk 4]
```diff
@@ -279,27 +304,66 @@ class FetchAuthSessionStateMachine extends FetchAuthSessionStateMachineBase {
...
+          identityIdResult = CognitoAuthSessionResult.error(e);
+        }
+      }
+    } else if (hasIdentityPool) {
+      credentialsResult = CognitoAuthSessionResult.success(awsCredentials);
+      identityIdResult = CognitoAuthSessionResult.success(
+        result.data.identityId,
+      );
+    } else {
+      credentialsResult = const CognitoAuthSessionResult<AWSCredentials>.error(
```
Comment (dnys1) [inline packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/fetch_auth_session_state_machine.dart:352]: Can we exit early with this before the try/catch?

File: packages/auth/amplify_auth_cognito_dart/lib/src/model/session/cognito_auth_session.dart [hunk 5]
```diff
@@ -18,27 +19,36 @@ class CognitoAuthSession extends AuthSession
...
 
   /// The user ID (subject).
-  final String? userSub;
+  String? get userSub {
+    try {
+      return _userPoolTokensResult?.value?.userId;
+    } on Object {
+      // If there was an exception fetching userPoolTokens, return the cached
+      // value
+      return _userPoolTokensResult?.previousValue?.userId;
```
[the remark this answers is not in the document]
Comment (dnys1) [reply packages/auth/amplify_auth_cognito_dart/lib/src/model/session/cognito_auth_session.dart:46]: Like you could have a `userIdResult` and do what you did for identityId in the state machine

**Review by dnys1 — COMMENTED** [review]
Looking good! Thanks for refactoring the fetch auth session tests!

File: packages/amplify_core/lib/src/types/exception/auth/signed_out_exception.dart [hunk 6]
```diff
@@ -15,6 +15,14 @@ class SignedOutException extends AuthException {
     super.underlyingException,
   });
 
+  /// Thrown when no user is currently signed in.
+  const SignedOutException.noUserSignedIn({
```
Comment (dnys1) [inline packages/amplify_core/lib/src/types/exception/auth/signed_out_exception.dart:19]: Should this be merged into the default constructor? I don't see why this case is special.

File: packages/auth/amplify_auth_cognito/lib/src/auth_plugin_impl.dart [hunk 7]
```diff
@@ -195,27 +195,27 @@ class _NativeAmplifyAuthCognito
...
     try {
-      final authSession = await _basePlugin.fetchAuthSession(
-        options: CognitoSessionOptions(getAWSCredentials: getAwsCredentials),
-      );
+      final authSession = await _basePlugin.fetchAuthSession();
       final nativeAuthSession = NativeAuthSession(
         isSignedIn: authSession.isSignedIn,
-        userSub: authSession.userSub,
-        identityId: authSession.identityId,
+        userSub: _valueOrNull(() => authSession.userSubResult.value),
```
Comment (dnys1) [inline packages/auth/amplify_auth_cognito/lib/src/auth_plugin_impl.dart:203]: Let's add `T? get valueOrNull` to `AWSResult`. This seems like something most people would want to do at one point or another.

File: packages/auth/amplify_auth_cognito_dart/lib/src/model/session/cognito_auth_session.dart [hunk 8]
```diff
@@ -2,52 +2,48 @@
...
 
-  /// {@macro amplify_auth_cognito.model.cognito_auth_session}
-  factory CognitoAuthSession.fromJson(Map<String, Object?> json) =>
-      _$CognitoAuthSessionFromJson(json);
-
-  /// The AWS credentials.
-  final AWSCredentials? credentials;
-
   /// The User Pool tokens.
-  final CognitoUserPoolTokens? userPoolTokens;
```
Comment (dnys1) [inline packages/auth/amplify_auth_cognito_dart/lib/src/model/session/cognito_auth_session.dart:35]: Can we leave these, and mark them as deprecated to a make this relatively non-breaking?

```dart
@Deprecated('Use `userPoolTokensResult.value` instead')
CognitoUserPoolTokens? get userPoolTokens => userPoolTokensResult.valueOrNull;
```

File: packages/auth/amplify_auth_cognito_dart/lib/src/model/session/cognito_auth_session.dart [hunk 9]
```diff
@@ -2,52 +2,48 @@
...
-        identityId,
+        userPoolTokensResult,
+        userSubResult,
+        credentialsResult,
+        identityIdResult,
       ];
 
   @override
-  Map<String, Object?> toJson() => _$CognitoAuthSessionToJson(this);
+  Map<String, Object?> toJson() => <String, dynamic>{
```
Comment (dnys1) [inline packages/auth/amplify_auth_cognito_dart/lib/src/model/session/cognito_auth_session.dart:46]: `toJson` should include all the results' presence IMO.

```json5
{
  'isSignedIn': false,
  'userPoolTokens': null,
  'awsCredentials': 'Instance of AWSCredentials',
  // etc.
}
```

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/fetch_auth_session_state_machine.dart [hunk 10]
```diff
@@ -239,22 +265,45 @@ class FetchAuthSessionStateMachine extends FetchAuthSessionStateMachineBase {
...
+      credentialsResult = AuthResult.error(e);
+      identityIdResult = AuthResult.error(e);
+    } on Object catch (e) {
+      final authException = UnknownException(
+        'An unknown exception occurred.',
+        underlyingException: e,
+      );
+      credentialsResult = AuthResult.error(authException);
+      identityIdResult = AuthResult.error(authException);
+    }
```
Comment (dnys1) [inline packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/fetch_auth_session_state_machine.dart:293]: Catching `Object` is risky in general since we may swallow errors that are not expected, like incorrect code. At the state machine layer and above, I'm not sure it's every really needed.

```suggestion
    } on Exception catch (e) {
      final authException = AuthException.fromException(e);
      credentialsResult = AuthResult.error(authException);
      identityIdResult = AuthResult.error(authException);
    }
```

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/fetch_auth_session_state_machine.dart [hunk 11]
```diff
@@ -279,27 +331,109 @@ class FetchAuthSessionStateMachine extends FetchAuthSessionStateMachineBase {
...
       }
-      userPoolTokens = await _refreshUserPoolTokens(userPoolTokens);
+      try {
+        userPoolTokens = await _refreshUserPoolTokens(userPoolTokens);
+        userPoolTokensResult = AuthResult.success(userPoolTokens);
+        userSubResult = AuthResult.success(userPoolTokens.userId);
+      } on AuthException catch (e) {
+        userPoolTokensResult = AuthResult.error(e);
+        userSubResult = AuthResult.error(e);
+      } on AWSHttpException catch (e) {
```
Comment (dnys1) [inline packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/fetch_auth_session_state_machine.dart:341]: Change as above and let's add this case to `AuthException.fromException`

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/fetch_auth_session_state_machine.dart [hunk 12]
```diff
@@ -279,27 +331,109 @@ class FetchAuthSessionStateMachine extends FetchAuthSessionStateMachineBase {
...
+              idToken == null ? null : _FederatedIdentity.cognito(idToken),
+        );
+        identityId = awsCredentialsResult.identityId;
+        awsCredentials = awsCredentialsResult.awsCredentials;
+        credentialsResult = AuthResult.success(awsCredentials);
+        identityIdResult = AuthResult.success(identityId);
+      } on AuthException catch (e) {
+        credentialsResult = AuthResult.error(e);
+        identityIdResult = AuthResult.error(e);
+      } on AWSHttpException catch (e) {
```
Comment (dnys1) [inline packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/fetch_auth_session_state_machine.dart:397]: Change as above

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/fetch_auth_session_state_machine.dart [hunk 13]
```diff
@@ -279,27 +331,109 @@ class FetchAuthSessionStateMachine extends FetchAuthSessionStateMachineBase {
...
+        );
+        credentialsResult = AuthResult.error(authException);
+        identityIdResult = AuthResult.error(authException);
+      }
+    } else {
+      credentialsResult = awsCredentials != null
+          ? AuthResult.success(awsCredentials)
+          : const AuthResult.error(
+              UnknownException(
+                'awsCredentials is null, but refreshAwsCredentials was false.',
```
Comment (dnys1) [inline packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/fetch_auth_session_state_machine.dart:417]: Applies to all instances of `UnknownException` in this file: Let's force unwrap unless we can think of a case where this would be true and prove it through a test.

File: packages/auth/amplify_auth_cognito_test/test/credentials/auth_plugin_credentials_provider_test.dart [hunk 14]
```diff
@@ -19,16 +19,6 @@ void main() {
     late AuthPluginCredentialsProviderImpl provider;
     late CognitoAuthStateMachine stateMachine;
 
-    // Performs the initial fetch so that credentials cache is hydrated.
-    Future<void> fetchAuthSession() async {
```
Comment (dnys1) [inline packages/auth/amplify_auth_cognito_test/test/credentials/auth_plugin_credentials_provider_test.dart:23]: Why was this removed?

File: packages/authenticator/amplify_authenticator/lib/src/services/amplify_auth_service.dart [hunk 15]
```diff
@@ -162,12 +162,20 @@ class AmplifyAuthService implements AuthService {
...
+      return true;
     } on SignedOutException {
       return false;
-    } on Exception {
+    } on NetworkException {
+      // NetworkException indicates that access and/or id tokens have expired
+      // and cannot be refreshed due to a network error. In this case the user
+      // should be treated as authenticated to allow for offline use cases.
+      return true;
+    } on Object {
```
Comment (dnys1) [inline packages/authenticator/amplify_authenticator/lib/src/services/amplify_auth_service.dart:177]:
```suggestion
    } on Exception {
```

File: packages/aws_common/lib/src/result/aws_result.dart [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+
+  /// The value of the result, if the result was successful.
+  ///
+  /// If an exception was thrown while retrieving the value, this will throw.
+  /// See [exception] for more details.
+  T get value {
+    switch (type) {
+      case AWSResultType.success:
+        if (_value == null) {
+          // this should not happen since AWSResultType.success requires a non
```
Comment (dnys1) [inline packages/aws_common/lib/src/result/aws_result.dart:47]: Well, it literally cannot happen - safe to unwrap here since it's guaranteed by the compiler.

File: packages/aws_common/lib/src/result/aws_result.dart [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+          // null value.
+          throw StateError(
+            'null value was provided for a successful AuthResult.',
+          );
+        }
+        return _value!;
+
+      case AWSResultType.error:
+        // ignore: only_throw_errors
+        throw exception!;
```
Comment (dnys1) [inline packages/aws_common/lib/src/result/aws_result.dart:57]:
```suggestion
        if (stackTrace != null) {
          // TODO(dnys1): Chain, instead, so that the current stack trace can provide context to the original?
        }
        throw exception!;
```

**Review by dnys1 — APPROVED** [review]
Looks good, thanks!

File: packages/auth/amplify_auth_cognito_dart/lib/src/model/session/cognito_session_options.dart [hunk 18]
```diff
@@ -13,25 +13,15 @@ class CognitoSessionOptions extends AuthSessionOptions
     with AWSEquatable<CognitoSessionOptions>, AWSDebuggable {
   /// {@macro amplify_auth_cognito.model.cognito_session_options}
   const CognitoSessionOptions({
-    this.getAWSCredentials = false,
```
Comment (dnys1) [inline packages/auth/amplify_auth_cognito_dart/lib/src/model/session/cognito_session_options.dart:16]: Thinking this over some more. Can we leave this in the constructor and mark it as deprecated with directions to just remove it, that it no longer does anything?

I'm worried a bit about the number of breaking changes and how this one, in particular, could be confusing.

File: packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/fetch_auth_session_state_machine.dart (hunk 10 above)
Comment (ragingsquirrel3) [reply packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/fetch_auth_session_state_machine.dart:293]: +1

**Review by ragingsquirrel3 — APPROVED** [review]

File: packages/auth/amplify_auth_cognito_test/test/credentials/auth_plugin_credentials_provider_test.dart (hunk 14 above)
Comment (dnys1) [reply packages/auth/amplify_auth_cognito_test/test/credentials/auth_plugin_credentials_provider_test.dart:23]: Ah gotcha - thanks!
